async.go

```diff
@@ -3,25 +3,140 @@ package async
 import (
 	"context"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/netbookai/log"
 	"github.com/netbookai/log/loggers"
 	"github.com/pkg/errors"
 )
 
-func run(ctx context.Context, timeout time.Duration, fn func(context.Context)) error {
-	errCh := make(chan error)
+type taskGroup struct {
+	tasks []task
+}
+
+type task struct {
+	name        string
+	taskWrapper func(context.Context) error
```

Comment:
could be just `fn`, task is already the name of the struct so extra context is not needed.